includes/class-amp-story-post-type.php

```diff
@@ -1490,6 +1522,232 @@ public static function insert_attachment( $object, $cropped ) {
+					if ( false !== $contents ) {
+						$zip->addFromString( $slug . '/assets/' . basename( $asset ), $contents );
+					}
+				}
+			}
+
+			// Close the active archive.
+			$zip->close();
+
+			// Read the file.
+			$fo = @fopen( $file, 'r' ); // phpcs:ignore
+
+			if ( ! $fo ) {
+				return new WP_Error( 'amp_story_export_file_open', esc_html__( 'Could not open the generated ZIP archive.', 'amp' ) );
+			} else {
+				header( 'Content-type: application/octet-stream' );
+				header( 'Content-Disposition: attachment; filename="' . $file . '"' );
+				header( 'Content-length: ' . filesize( $file ) );
+				fpassthru( $fo );
+				unlink( $file );
+				die();
+			}
+		} else {
+			switch ( $res ) {
+				case ZipArchive::ER_EXISTS:
+					$msg = esc_html__( 'Zip archive already exists.', 'amp' );
+					break;
+				case ZipArchive::ER_INCONS:
+					$msg = esc_html__( 'Zip archive inconsistent.', 'amp' );
+					break;
+				case ZipArchive::ER_MEMORY:
+					$msg = esc_html__( 'Malloc failure.', 'amp' );
+					break;
+				case ZipArchive::ER_NOENT:
+					$msg = esc_html__( 'No such zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_NOZIP:
+					$msg = esc_html__( 'Not a zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_OPEN:
+					$msg = esc_html__( 'Could not open zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_READ:
+					$msg = esc_html__( 'Zip archive read error.', 'amp' );
+					break;
+				case ZipArchive::ER_SEEK:
+					$msg = esc_html__( 'Zip archive seek error.', 'amp' );
+					break;
+				default:
+					$msg = esc_html__( 'Unknown zip archive error', 'amp' );
+					break;
+			}
+			return new WP_Error( 'amp_story_zip_archive_error', $msg );
+		}
+	}
+
+	/**
+	 * Ensures we can preview the story even when it's a draft.
+	 *
+	 * @param WP_Post[] $posts Array of post objects.
+	 * @param WP_Query  $query The WP_Query instance (passed by reference).
+	 *
+	 * @return WP_Post[] The post results array.
+	 */
+	public static function enable_export_preview( $posts, $query ) {
+		if ( 1 !== count( $posts ) ) {
+			return $posts;
+		}
+
+		$is_preview = (
+			$query->is_preview
+			&& self::POST_TYPE_SLUG === $posts[0]->post_type
+			&& isset( $_GET['story_export'] ) // phpcs:ignore
+			&& isset( $_GET['_wpnonce'] ) // phpcs:ignore
+			// phpcs:ignore && wp_verify_nonce( wp_unslash( $_GET['_wpnonce'] ), 'amp-story-export-' . $posts[0]->ID )
+		);
+
+		if ( $is_preview ) {
+			$query->_story_export = $posts;
+			add_filter( 'the_posts', array( __CLASS__, 'display_export_preview' ), 10, 2 );
+		}
+
+		return $posts;
+	}
+
+	/**
+	 * Replaces the array of retrieved posts after they've been fetched and
+	 * internally processed with the exported story.
+	 *
+	 * @param WP_Post[] $posts Array of post objects.
+	 * @param WP_Query  $query The WP_Query instance (passed by reference).
+	 *
+	 * @return WP_Post[] The post results array.
+	 */
+	public static function display_export_preview( $posts, $query ) {
```

Comment:
Instead of this I suggest passing the user's `$_COOKIE` through in the call to `wp_remote_get()`